### PR TITLE
tsify CellAction, CellRowDeleteButton and CellExpander

### PR DESCRIPTION
**Pull request by amanmahajan7**
Also cleaned up Cell component.

### CODE REVIEW

File: packages/react-data-grid/src/Cell.js [hunk 1]
```diff
@@ -286,76 +271,84 @@ export default class Cell extends React.PureComponent {
     this.node = node;
   };
 
+  renderCellContent() {
+    let cellContent;
+    const { value, column, height, tooltip, isScrolling, isExpanded, expandableOptions, cellMetaData } = this.props;
+    const Formatter = column.formatter;
+    const cellProps = {
```
Comment (amanmahajan7) [inline packages/react-data-grid/src/Cell.js]: This ensures we pass the same props to components and elements

File: packages/react-data-grid/src/CellAction.tsx [hunk 2]
```diff
@@ -1,16 +1,23 @@
 import classNames from 'classnames';
-import React from 'react';
-import PropTypes from 'prop-types';
+import React, { ReactElement } from 'react';
 
-import CellActionShape from './common/prop-shapes/CellActionShape';
+interface Action {
+  icon: string | ReactElement;
+  actions: Array<{ text: string; callback: () => void }>;
```
Comment (qili26) [inline packages/react-data-grid/src/CellAction.tsx]: maybe add an interface for `{ text: string; callback: () => void }`?

File: packages/react-data-grid/src/Cell.js [hunk 3]
```diff
@@ -68,123 +67,109 @@ export default class Cell extends React.PureComponent {
...
   };
 
-  onDragOver = (e) => {
+  handleDragOver = (e) => {
     e.preventDefault();
   };
 
-  getStyle = () => {
-    const style = {
+  getStyle() {
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply packages/react-data-grid/src/Cell.js]: `getStyle` is called inside the render function so `this` will always be correct and we do not need to bind it

File: packages/react-data-grid/src/Cell.js [hunk 4]
```diff
@@ -68,123 +67,109 @@ export default class Cell extends React.PureComponent {
...
   getRowData = (props = this.props) => {
     return props.rowData.toJSON ? props.rowData.toJSON() : props.rowData;
   };
 
-  getFormatterDependencies = () => {
+  getFormatterDependencies() {
     // convention based method to get corresponding Id or Name of any Name or Id property
-    if (typeof this.props.column.getRowMetaData === 'function') {
+    const { getRowMetaData } = this.props.column;
+    if (typeof getRowMetaData === 'function') {
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply packages/react-data-grid/src/Cell.js]: I am thinking of removing `isFunction` but for the time being I can use `isFunction` for consistency

File: packages/react-data-grid/src/Cell.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply packages/react-data-grid/src/Cell.js]: Let's tackle this in a separate PR when we migrate `Cell` to typescript. We can probably add `BaseFormatterProps`

File: packages/react-data-grid/src/Cell.js [hunk 5]
```diff
@@ -68,123 +67,109 @@ export default class Cell extends React.PureComponent {
...
     const meta = this.props.cellMetaData;
     if (meta != null && meta.onCellExpand != null) {
-      meta.onCellExpand({ rowIdx: this.props.rowIdx, idx: this.props.idx, rowData: this.props.rowData, expandArgs: this.props.expandableOptions });
-    }
-  };
-
-  onCellKeyDown = (e) => {
-    if (this.canExpand() && e.key === 'Enter') {
-      this.onCellExpand(e);
+      meta.onCellExpand({
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply packages/react-data-grid/src/Cell.js]: Added `getSubRowOptions`

File: packages/react-data-grid/src/CellAction.tsx (hunk 2 above)
Comment (amanmahajan7) [reply packages/react-data-grid/src/CellAction.tsx]: Done

File: packages/react-data-grid/src/Cell.js [hunk 6]
```diff
@@ -286,76 +271,84 @@ export default class Cell extends React.PureComponent {
...
+        expandableOptions={this.props.expandableOptions}
+        onCellExpand={this.handleCellExpand}
+      />
+    );
 
-    const cellExpander = this.canExpand()
-      && <CellExpand expandableOptions={this.props.expandableOptions} onCellExpand={this.onCellExpand} />;
     return (
       <div
+        ref={this.setCellRef}
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply packages/react-data-grid/src/Cell.js]: Done

File: packages/react-data-grid/src/Cell.js [hunk 7]
```diff
@@ -247,115 +223,119 @@ export default class Cell extends React.PureComponent {
...
 
-  renderCellContent = (props) => {
-    let CellContent;
-    const Formatter = this.getFormatter();
     if (isElement(Formatter)) {
-      CellContent = React.cloneElement(Formatter, { ...props, dependentValues: this.getFormatterDependencies(), row: this.getRowData() });
+      cellContent = React.cloneElement(Formatter, cellProps);
     } else if (isValidElementType(Formatter)) {
-      CellContent = <Formatter value={this.props.value} dependentValues={this.getFormatterDependencies()} isScrolling={this.props.isScrolling} row={this.getRowData()} />;
+      cellContent = <Formatter {...cellProps} />;
```
Comment (MayhemYDG) [inline packages/react-data-grid/src/Cell.js]: You can do `React.createElement(column.formatter, cellProps)`, that way you don't need to re-assign `column.formatter` just do to valid jsx.
Bonus point for looking similar to the `React.cloneElement` line. 🙂

File: packages/react-data-grid/src/CellAction.tsx (hunk 2 above)
Comment (qili26) [reply packages/react-data-grid/src/CellAction.tsx]: sure.

File: packages/react-data-grid/src/CellExpander.tsx [hunk 8]
```diff
@@ -1,30 +1,32 @@
...
 
 import { CellExpand } from './common/constants';
 
-export default class CellExpander extends React.Component {
-  static propTypes = {
-    expandableOptions: PropTypes.object.isRequired,
-    onCellExpand: PropTypes.func.isRequired
-  };
+export interface CellExpanderProps {
+  expandableOptions?: { expanded?: boolean };
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply packages/react-data-grid/src/CellExpander.tsx]: Done

File: packages/react-data-grid/src/CellAction.tsx [hunk 9]
```diff
@@ -1,16 +1,23 @@
 import classNames from 'classnames';
-import React from 'react';
-import PropTypes from 'prop-types';
+import React, { ReactElement } from 'react';
 
-import CellActionShape from './common/prop-shapes/CellActionShape';
+interface Action {
+  icon: string | ReactElement;
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply packages/react-data-grid/src/CellAction.tsx]: I changed it to `ReactNode` so users can specify whatever they want. I am also thinking of cleaning up cell actions logic somehow so users can specify their own dropdown.

File: packages/react-data-grid/src/CellExpander.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (MayhemYDG) [reply packages/react-data-grid/src/CellExpander.tsx]: Looks a lot better. 🙂

File: packages/react-data-grid/src/__tests__/CellAction.spec.tsx [hunk 10]
```diff
@@ -1,17 +1,17 @@
...
+const setup = (overriderProps: Partial<CellActionProps> = {}) => {
   const props = {
     action: {
       icon: 'glyphicon glyphicon-link',
       callback: jest.fn()
     },
     isFirst: true,
     ...overriderProps
-  };
+  } as CellActionProps;
```
[the remark this answers is not in the document]
Comment (qili26) [reply packages/react-data-grid/src/__tests__/CellAction.spec.tsx]: @MayhemYDG 
A question to the test file. should we be as strict as in component files? My question is that for the unit testing, we might just need to focus on one function, it doesn't make sense for us to pass the whole data in the object. It's not like integration tests, we need mimic everything.

File: packages/react-data-grid/src/CellAction.tsx [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    }
+  }
+
+  return (
+    <div className={cellActionClasses} onMouseLeave={() => setIsOpen(false)}>
+      <div className={actionButtonClasses} onClick={onActionIconClick}>
+        {icon}
+      </div>
+      {
+        Array.isArray(actions) && actions.length && isOpen && (
```
Comment (MayhemYDG) [inline packages/react-data-grid/src/CellAction.tsx]: I think we should just trust our types. Here and above as well. WDYT?

```jsx
isOpen && actions.length > 0 && (
  ...
)
```

File: packages/react-data-grid/src/__tests__/CellAction.spec.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (MayhemYDG) [reply packages/react-data-grid/src/__tests__/CellAction.spec.tsx]: IMO we should have test helpers that generate well-formed, valid data, so we can re-use them across component tests without having to worry about missing values.

[the remark this answers is not in the document]
Comment (qili26) [reply packages/react-data-grid/src/__tests__/CellAction.spec.tsx]: I had searched that and we might not have this kind package anywhere. All test data creating package as like `fake address book` that creates the random data but it does't accept a type and then generate that data file.

[the remark this answers is not in the document]
Comment (MayhemYDG) [reply packages/react-data-grid/src/__tests__/CellAction.spec.tsx]: Yeah we'll need to make our own helpers.

File: packages/react-data-grid/src/CellAction.tsx (hunk 11 above)
Comment (qili26) [reply packages/react-data-grid/src/CellAction.tsx]: I would like to say we should keep it and clean step by step. The type is safe inside this file, but the `Cell.js` is still not `tsx` yet

Comment (MayhemYDG) [reply packages/react-data-grid/src/CellAction.tsx]: Let's hope we won't forget to clean it then. 🤞

Comment (amanmahajan7) [reply packages/react-data-grid/src/CellAction.tsx]: `actions` is not a required field so we cannot change it

Comment (amanmahajan7) [reply packages/react-data-grid/src/CellAction.tsx]: https://github.com/adazzle/react-data-grid/blob/next/examples/scripts/example27-cell-actions.js#L134

Comment (MayhemYDG) [reply packages/react-data-grid/src/CellAction.tsx]: Right, but `Array.isArray` is still overkill probably.

File: packages/react-data-grid/src/__tests__/CellAction.spec.tsx [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+      expect(actions[0].callback).not.toHaveBeenCalled();
+      renderedActionMenu.childAt(0).simulate('click');
+      expect(actions[0].callback).toHaveBeenCalled();
+      wrapper.unmount();
+    });
+  });
+
+  describe('when isFirst is passed', () => {
+    describe('when isFirst is true', () => {
+      xit('will render a button action with cell-action-last class on it', () => {
```
Comment (amanmahajan7) [inline packages/react-data-grid/src/__tests__/CellAction.spec.tsx]: This test is failing for some reason, may be something to do with hooks. I will fix it when I update `Cell` component

File: packages/react-data-grid/src/Cell.js [hunk 13]
```diff
@@ -68,168 +71,141 @@ export default class Cell extends React.PureComponent {
...
   }
 
   setScrollLeft = (scrollLeft) => {
-    const node = this.node;
+    const node = this.cell.current;
     if (node) {
-      const transform = `translate3d(${scrollLeft}px, 0px, 0px)`;
-      node.style.webkitTransform = transform;
-      node.style.transform = transform;
+      node.style.transform = `translate3d(${scrollLeft}px, 0px, 0px)`;
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply packages/react-data-grid/src/Cell.js]: Done

File: packages/react-data-grid/src/Cell.js (hunk 7 above)
Comment (amanmahajan7) [reply packages/react-data-grid/src/Cell.js]: I will leave it as it for now but I will revisit it when Cell is migrated to TS

**Pull request merged**